### PR TITLE
introduce a freelist interface

### LINKED ISSUE #773: Introduce freelist interface
bbolt features two types of freelist: "array" and "hashmap". They are currently implemented in one monolithic struct, with their differences in functions separated into two go files. 

Testing has been a pain so far, with many of the freelist details (i.e.releasing pages and serde to disk) leaking outside of the freelist struct. This makes reasoning about the implementations very difficult, too. Adding a new type of list is becoming increasingly cumbersome without a proper interface. 

Quickly pulling out the existing functions into an interface could look like this:

```
type Freelist interface {

	// Init initializes this freelist with the given list of pages. 
        // TODO currently is "readIDs", maybe should be called Read(page *common.Page)
	Init(ids common.Pgids) 

	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
	// If a contiguous block cannot be found then 0 is returned.
	Allocate(txid common.Txid, numPages int) common.Pgid

	// Count returns the number of free pages.
	Count() int

	// FreePageIds returns all free pages.
	FreePageIds() common.Pgids

	// MergeSpans merges the given pages into the existing freelist.
	MergeSpans(ids common.Pgids) // TODO should probably named "Release" or "Free"

	// Write writes the freelist into the given page. The freelist MUST fit into the given page.
	Write(page *common.Page) error
}
```

Ideally we would move this into its own internal package in order to avoid leaking any more internals in the future. HashMap and Array should be two implementation as its own struct. Some shared functions  (ie read/write) could be implemented in a common struct for now. Even though I expect those to become implementation dependent at some point, it seems quite inefficient to write low-digit uint64 integers without vint compression (or as a bitset), for example.

### PR DESCRIPTION
**Pull request by tjungblu**
This introduces an interface for the freelist, splits it into two concrete implementations.

fixes #773

### CODE REVIEW

File: db_test.go [hunk 1]
```diff
@@ -1053,8 +1053,8 @@ func TestDB_Stats(t *testing.T) {
 	stats := db.Stats()
 	if stats.TxStats.GetPageCount() != 2 {
 		t.Fatalf("unexpected TxStats.PageCount: %d", stats.TxStats.GetPageCount())
-	} else if stats.FreePageN != 0 {
-		t.Fatalf("unexpected FreePageN != 0: %d", stats.FreePageN)
+	} else if stats.FreePageN != 2 {
```
Comment (tjungblu) [inline db_test.go:1056]: I think I've messed up the stats, so expect this to go back to its original value

File: internal/freelist/freelist.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+// FreelistType is the type of the freelist backend
+type FreelistType string
```
Comment (tjungblu) [inline internal/freelist/freelist.go:10]: we might need to keep those in the bbolt package, or copy&paste them potentially

File: internal/freelist/freelist.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// PendingPageIds returns all pending pages by transaction id.
+	PendingPageIds() map[common.Txid]*txPending
```
Comment (tjungblu) [inline internal/freelist/freelist.go:44]: this could be private, only used in this package

**Comment by tjungblu** [comment]
rebased to the latest freelist refactoring

File: internal/freelist/freelist.go (hunk 2 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:10]: moved it back to the bbolt package

File: internal/freelist/freelist.go (hunk 3 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:44]: made it private now

**Comment by tjungblu** [comment]
let's see how the benchmarks perform :)

File: internal/freelist/hashmap.go [hunk 4]
```diff
@@ -8,26 +8,57 @@ import (
...
-	common.Verify(func() {
-		expectedFreePageCount := f.hashmapFreeCountSlow()
-		common.Assert(int(f.freePagesCount) == expectedFreePageCount,
-			"freePagesCount (%d) is out of sync with free pages map (%d)", f.freePagesCount, expectedFreePageCount)
-	})
-	return int(f.freePagesCount)
+// pidSet holds the set of starting pgids which have the same span size
+type pidSet map[common.Pgid]struct{}
+
+type HashMap struct {
```
Comment (ahrtr) [inline internal/freelist/hashmap.go:14]:
```suggestion
type hashmap struct {
```

File: internal/freelist/array.go [hunk 5]
```diff
@@ -7,14 +7,18 @@ import (
 	"go.etcd.io/bbolt/internal/common"
 )
 
-// arrayFreeCount returns count of free pages(array version)
-func (f *freelist) arrayFreeCount() int {
-	return len(f.ids)
+type Array struct {
```
Comment (ahrtr) [inline internal/freelist/array.go:10]:
```suggestion
type array struct {
```

File: internal/freelist/serde.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+type Serializable interface {
+	// Read calls Init with the page ids stored in te given page.
+	Read(f Freelist, page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(f Freelist, page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the page after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize(f Freelist) int
```
Comment (ahrtr) [inline internal/freelist/serde.go:20]: Let's try not add something new in this PR to make the review easier. Anything new should be discussed separately
```suggestion
```

File: internal/freelist/freelist.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type Freelist interface {
```
Comment (ahrtr) [inline internal/freelist/freelist.go:9]: Refer to https://github.com/golang/go/blob/ea537cca314d9da5365eeefcc375410c76e93b36/src/sort/sort.go#L14
```suggestion
type Interface interface {
```

File: internal/freelist/freelist.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	Free(txId common.Txid, p *common.Page)
+
+	// Freed returns whether a given page is in the free list.
+	Freed(pgId common.Pgid) bool
+
+	// Rollback removes the pages from a given pending tx.
+	Rollback(txId common.Txid)
+
+	// pendingPageIds returns all pending pages by transaction id.
+	pendingPageIds() map[common.Txid]*txPending
```
Comment (ahrtr) [inline internal/freelist/freelist.go:46]: It's just an internal method, no need to add it into the interface.
```suggestion
```

File: internal/freelist/hashmap.go [hunk 9]
```diff
@@ -74,17 +105,21 @@ func (f *freelist) hashmapAllocate(txid common.Txid, n int) common.Pgid {
 	return 0
 }
 
-// hashmapReadIDs reads pgids as input an initial the freelist(hashmap version)
-func (f *freelist) hashmapReadIDs(pgids []common.Pgid) {
-	f.init(pgids)
+func (f *HashMap) Count() int {
+	return f.FreeCount() + f.PendingCount()
+}
```
Comment (ahrtr) [inline internal/freelist/hashmap.go:110]: It has identical implementation as array.Count(), so let's move it into shared?

File: internal/freelist/array.go [hunk 10]
```diff
@@ -96,3 +101,16 @@ func (f *freelist) arrayMergeSpans(ids common.Pgids) {
 	})
 	f.ids = common.Pgids(f.ids).Merge(ids)
 }
+
+func NewArray() *Array {
```
Comment (ahrtr) [inline internal/freelist/array.go:105]:
```suggestion
func NewArrayFreelist() *Array {
```

File: internal/freelist/hashmap.go [hunk 11]
```diff
@@ -279,3 +283,19 @@ func (f *freelist) idsFromBackwardMap() map[common.Pgid]struct{} {
 	}
 	return ids
 }
+
+func NewHashMap() *HashMap {
```
Comment (ahrtr) [inline internal/freelist/hashmap.go:287]:
```suggestion
func NewHashMapFreelist() *HashMap {
```

File: internal/freelist/freelist.go (hunk 2 above)
Comment (ahrtr) [reply internal/freelist/freelist.go]: Right, we can't break the user experience in 1.4.

File: internal/freelist/hashmap.go (hunk 9 above)
Comment (tjungblu) [reply internal/freelist/hashmap.go:110]: I wish, `FreeCount()` is implementation dependent, so we can't move it into shared. We can deal with it the same way mergeSpans works however

File: internal/freelist/freelist.go (hunk 8 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:46]: you need it in the interface for Copyall & NoSyncReload

File: internal/freelist/freelist.go (hunk 7 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:9]: done

File: internal/freelist/array.go (hunk 5 above)
Comment (tjungblu) [reply internal/freelist/array.go:10]: done

File: internal/freelist/hashmap.go (hunk 4 above)
Comment (tjungblu) [reply internal/freelist/hashmap.go:14]: done

File: internal/freelist/array.go (hunk 10 above)
Comment (tjungblu) [reply internal/freelist/array.go:105]: done

File: internal/freelist/hashmap.go (hunk 11 above)
Comment (tjungblu) [reply internal/freelist/hashmap.go:287]: done

File: internal/freelist/serde.go [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	// Read calls Init with the page ids stored in te given page.
+	Read(f Interface, page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(f Interface, page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize(f Interface) int
+}
```
Comment (ahrtr) [inline internal/freelist/serde.go:21]: Propose to change the interface something like below,
```
type ReadWriter interface {
	// Read initializes the freelist from the given page.
	Read(page *common.Page)

	// Write writes the freelist into the given page.
	Write(page *common.Page)

	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
	// This should never underestimate the size.
	EstimatedWritePageSize() int
}
```

Also embedded it into Interface,
```
type Interface interface {
	ReadWriter
	...... 
}
```

Comment (tjungblu) [reply internal/freelist/serde.go:21]: +1 on the naming, the pull into the interface has some coupling issues though. When we move it into `shared` (which would make sense), the following functions are missing from the shared struct: `Count`, `Init` and `CopyAll` would also need to become part of the implementation in shared (which again depends on `PendingCount` and `FreePageIds`).

I think ultimately `shared` will just become the old freelist.go if we continue down that path ;-)

Otherwise we can move the three read/write methods copied into both array+hashmap, which also resolves it.

WDYT?

Comment (tjungblu) [reply internal/freelist/serde.go:21]: also, maybe a more general question: should the serialization be implementation dependent or interface-dependent? 

Currently we have to keep the serde the exact same between array/hashmap, otherwise we won't be able to switch the freelist implementation "at runtime" after a bbolt has been created. If we make it implementation-dependent, we need to devise a way to transform one to the other.

Comment (tjungblu) [reply internal/freelist/serde.go]: I've mangled it into the shared interface, similar to mergeSpans, added it into a separated commit for better comparison.

File: internal/freelist/freelist.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// Release moves all page ids for a transaction id (or older) to the freelist.
+	Release(txId common.Txid)
+
+	// ReleaseRange moves pending pages allocated within an extent [begin,end] to the free list.
+	ReleaseRange(begin, end common.Txid)
```
Comment (ahrtr) [inline internal/freelist/freelist.go:47]: Once https://github.com/etcd-io/bbolt/pull/783 gets merged, then no need to export such interface methods.

File: internal/freelist/serde.go (hunk 12 above)
Comment (tjungblu) [reply internal/freelist/serde.go]: I've saved the commit into another branch https://github.com/tjungblu/bbolt/tree/refactor_freelists_commits

squashed this one here to have an easier time rebasing on top of #783

File: internal/freelist/hashmap.go (hunk 9 above)
Comment (tjungblu) [reply internal/freelist/hashmap.go]: added it to the shared interface

File: internal/freelist/freelist.go (hunk 13 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:47]: put it into shared, but keeping it package private for the unit tests that still depend on this

File: internal/freelist/freelist.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
```
Comment (tjungblu) [inline internal/freelist/freelist.go:38]: that can be package private now, used only by ReadWriter and unit tests

File: internal/freelist/freelist.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	PendingCount() int
+
+	// AddReadonlyTXID adds a given read-only transaction id for pending page tracking.
+	AddReadonlyTXID(txid common.Txid)
+
+	// RemoveReadonlyTXID removes a given read-only transaction id from pending page tracking.
+	RemoveReadonlyTXID(txid common.Txid)
+
+	// FreePages releases any pages associated with closed read-only transactions.
+	FreePages()
```
Comment (tjungblu) [inline internal/freelist/freelist.go:50]: maybe to stay consistent with the other R-TX?

```suggestion
	FreeReadonlyTxPages()
```

File: internal/freelist/shared.go [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+	Init(ids common.Pgids)
+	// Count returns the number of free and pending pages.
+	Count() int
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+	// mergeSpans is merging the given pages into the freelist
+	mergeSpans(ids common.Pgids)
+}
```
Comment (ahrtr) [inline internal/freelist/shared.go:23]: You can add `mergeSpans` into `Interface`
```suggestion
```

File: internal/freelist/shared.go [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+
+type txPending struct {
+	ids              []common.Pgid
+	alloctx          []common.Txid // txids allocating the ids
+	lastReleaseBegin common.Txid   // beginning txid of last matching releaseRange
+}
+
+type shared struct {
+	sharedInterface
```
Comment (ahrtr) [inline internal/freelist/shared.go:32]:
```suggestion
	Interface
```

File: internal/freelist/shared.go [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+			// Pending free aborted; restore page back to alloc list.
+			t.allocs[pgid] = tx
+		} else {
+			// Freed page was allocated by this txn; OK to throw away.
+			m = append(m, pgid)
+		}
+	}
+	// Remove pages from pending list and mark as free if allocated by txid.
+	delete(t.pending, txid)
+	t.sharedInterface.mergeSpans(m)
```
Comment (ahrtr) [inline internal/freelist/shared.go:121]:
```suggestion
	t.mergeSpans(m)
```

File: internal/freelist/freelist.go [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// Reload reads the freelist from a page and filters out pending items.
+func Reload(f Interface, p *common.Page) {
+	f.Read(p)
+	NoSyncReload(f, p.FreelistPageIds())
+}
+
+// NoSyncReload reads the freelist from Pgids and filters out pending items.
+func NoSyncReload(f Interface, pgIds common.Pgids) {
```
Comment (ahrtr) [inline internal/freelist/freelist.go:90]: Can we turn these three functions into methods? Also Copyall is identical to `(t *shared) copyall`

File: internal/freelist/array.go [hunk 20]
```diff
@@ -96,3 +97,16 @@ func (f *freelist) arrayMergeSpans(ids common.Pgids) {
...
+func NewArrayFreelist() Interface {
+	a := &array{
+		shared: shared{
+			pending: make(map[common.Txid]*txPending),
+			allocs:  make(map[common.Pgid]common.Txid),
+			cache:   make(map[common.Pgid]struct{}),
+		},
+	}
+	// this loopy reference allows us to share the span merging via interfaces
+	a.shared.sharedInterface = a
```
Comment (ahrtr) [inline internal/freelist/array.go:110]:
```suggestion
	a.shared.Interface = a
```

File: internal/freelist/hashmap.go [hunk 21]
```diff
@@ -279,3 +279,19 @@ func (f *freelist) idsFromBackwardMap() map[common.Pgid]struct{} {
...
+			allocs:  make(map[common.Pgid]common.Txid),
+			cache:   make(map[common.Pgid]struct{}),
+			pending: make(map[common.Txid]*txPending),
+		},
+		freemaps:    make(map[uint64]pidSet),
+		forwardMap:  make(map[common.Pgid]uint64),
+		backwardMap: make(map[common.Pgid]uint64),
+	}
+	// this loopy reference allows us to share the span merging via interfaces
+	hm.shared.sharedInterface = hm
```
Comment (ahrtr) [inline internal/freelist/hashmap.go:295]:
```suggestion
	hm.shared.Interface = hm
```

File: internal/freelist/array.go [hunk 22]
```diff
@@ -7,14 +7,18 @@ import (
...
-	return len(f.ids)
+type array struct {
+	shared
+
+	ids []common.Pgid // all free and available free page ids.
+}
+
+func (f *array) Init(ids common.Pgids) {
+	f.ids = ids
+	f.reindex(f.FreePageIds(), f.pendingPageIds())
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/array.go:18]: yep, done

File: internal/freelist/freelist.go [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+type Interface interface {
+	ReadWriter
+
+	// Init initializes this freelist with the given list of pages.
+	Init(ids common.Pgids)
+
+	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
+	// If a contiguous block cannot be found then 0 is returned.
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/freelist.go:28]: added

File: internal/freelist/freelist.go [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/freelist.go:37]: added

File: internal/freelist/freelist.go [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// AddReadonlyTXID adds a given read-only transaction id for pending page tracking.
+	AddReadonlyTXID(txid common.Txid)
+
+	// RemoveReadonlyTXID removes a given read-only transaction id from pending page tracking.
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/freelist.go:46]: added

File: internal/freelist/freelist.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/freelist.go:50]: updated

File: internal/freelist/shared.go (hunk 16 above)
Comment (tjungblu) [reply internal/freelist/shared.go:23]: done

File: internal/freelist/shared.go (hunk 17 above)
Comment (tjungblu) [reply internal/freelist/shared.go:32]: done

File: internal/freelist/freelist.go (hunk 19 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:90]: I found those easier to grasp with composition, but I can add those back into the shared interface as with copyall

File: internal/freelist/array.go (hunk 20 above)
Comment (tjungblu) [reply internal/freelist/array.go:110]: done

File: internal/freelist/hashmap.go (hunk 21 above)
Comment (tjungblu) [reply internal/freelist/hashmap.go:295]: done

File: internal/freelist/freelist.go (hunk 19 above)
Comment (tjungblu) [reply internal/freelist/freelist.go:90]: moved them back, replaced it with their old exact implementation in freelist.go

File: internal/freelist/array.go [hunk 26]
```diff
@@ -7,14 +7,18 @@ import (
 	"go.etcd.io/bbolt/internal/common"
 )
 
-// arrayFreeCount returns count of free pages(array version)
-func (f *freelist) arrayFreeCount() int {
-	return len(f.ids)
+type array struct {
+	shared
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/array.go:11]: applied

File: internal/freelist/hashmap.go [hunk 27]
```diff
@@ -8,26 +8,57 @@ import (
...
-		expectedFreePageCount := f.hashmapFreeCountSlow()
-		common.Assert(int(f.freePagesCount) == expectedFreePageCount,
-			"freePagesCount (%d) is out of sync with free pages map (%d)", f.freePagesCount, expectedFreePageCount)
-	})
-	return int(f.freePagesCount)
+// pidSet holds the set of starting pgids which have the same span size
+type pidSet map[common.Pgid]struct{}
+
+type hashMap struct {
+	shared
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/hashmap.go:15]: applied

File: internal/freelist/shared.go [hunk 28]
```diff
@@ -0,0 +1,313 @@
...
+
+type shared struct {
+	Interface
+
+	readonlyTXIDs []common.Txid               // all readonly transaction IDs.
+	allocs        map[common.Pgid]common.Txid // mapping of Txid that allocated a pgid.
+	cache         map[common.Pgid]struct{}    // fast lookup of all free and pending page ids.
+	pending       map[common.Txid]*txPending  // mapping of soon-to-be free page ids by tx.
+}
+
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/shared.go:26]: applied

File: internal/freelist/array.go [hunk 29]
```diff
@@ -96,3 +97,16 @@ func (f *freelist) arrayMergeSpans(ids common.Pgids) {
...
 	f.ids = common.Pgids(f.ids).Merge(ids)
 }
+
+func NewArrayFreelist() Interface {
+	a := &array{
+		shared: shared{
+			pending: make(map[common.Txid]*txPending),
+			allocs:  make(map[common.Pgid]common.Txid),
+			cache:   make(map[common.Pgid]struct{}),
+		},
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/array.go:107]: applied

File: internal/freelist/hashmap.go [hunk 30]
```diff
@@ -279,3 +279,19 @@ func (f *freelist) idsFromBackwardMap() map[common.Pgid]struct{} {
...
 	return ids
 }
+
+func NewHashMapFreelist() Interface {
+	hm := &hashMap{
+		shared: shared{
+			allocs:  make(map[common.Pgid]common.Txid),
+			cache:   make(map[common.Pgid]struct{}),
+			pending: make(map[common.Txid]*txPending),
+		},
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply internal/freelist/hashmap.go:289]: applied

File: db.go [hunk 31]
```diff
@@ -418,12 +419,13 @@ func (db *DB) loadFreelist() {
...
-			db.freelist.readIDs(db.freepages())
+			db.freelist.Init(db.freepages())
 		} else {
 			// Read free list from freelist page.
-			db.freelist.read(db.page(db.meta().Freelist()))
+			db.freelist.Read(db.page(db.meta().Freelist()))
 		}
-		db.stats.FreePageN = db.freelist.free_count()
+		db.stats.FreePageN = db.freelist.FreeCount()
+		db.stats.PendingPageN = db.freelist.PendingCount()
```
Comment (ahrtr) [inline db.go:428]: When loading freelist in the very first time, there isn't any TXNs yet, so the pending must be empty.
```suggestion
```

**Review by ahrtr — COMMENTED** [review]
The removed `freelist_test.go` has 485 lines of code, but the sum of the `array_test.go`, `freelist_test.go` and `hashmap_test.go` have around 360 lines of code, where is the remaining around 120 lines of test code?

File: internal/freelist/freelist.go (hunk 14 above)
Comment (ahrtr) [reply internal/freelist/freelist.go:37]: It seems true. We can just move it right above the `pendingPageIds` method.

**Comment by tjungblu** [comment]
> The removed freelist_test.go has 485 lines of code, but the sum of the array_test.go, freelist_test.go and hashmap_test.go have around 360 lines of code, where is the remaining around 120 lines of test code?

yep good catch, there were more hashmap tests at the bottom which are now back in action. Line counts are matching again?

File: db.go (hunk 31 above)
Comment (tjungblu) [reply db.go]: removed

**Comment by ahrtr** [comment]
> yep good catch, there were more hashmap tests at the bottom which are now back in action. Line counts are matching again?

Yes, makes sense now. thx.

One minor comment on the order of the methods in `shared.go`, please try to reorder the methods, e.g.
- put `release` and `releaseRange` right below `ReleasePendingPages`;
- put `Read`, `EstimatedWritePageSize` and `Write` at the bottom of the file;
- put `txIDx` and its implementation on sort.Interface right above `ReleasePendingPages`

**Comment by tjungblu** [comment]
done, PTAL again

**Review by ahrtr — APPROVED** [review]
Great work!  Thanks @tjungblu .

We can continue to think about how to add dedicated & randomized test case for the freelist management. We can also continue to refactor if needed.

cc @fuweid to take a look

**Comment by tjungblu** [comment]
Thanks for your patience! I think the best step forward is to test the interface for starters. Then we can check what else we can refactor from there...

**Comment by tjungblu** [comment]
@fuweid any comments?

File: internal/freelist/freelist_test.go [hunk 32]
```diff
@@ -0,0 +1,282 @@
...
+
+	if got2 := f2.freePageIds(); !reflect.DeepEqual(got2, common.Pgids(exp2)) {
+		t.Fatalf("exp2=%#v; got2=%#v", exp2, got2)
+	}
+
+}
+
+// newTestFreelist get the freelist type from env and initial the freelist
+func newTestFreelist() Interface {
+	if env := os.Getenv(TestFreelistType); env == "map" {
```
Comment (tjungblu) [inline internal/freelist/freelist_test.go:277]: `
```suggestion
	if env := os.Getenv(TestFreelistType); env == "hashmap" {
```

that one was hiding the tests

**Comment by fuweid** [comment]
@tjungblu I will take a look in following two days. sorry for the delay.

File: internal/freelist/shared.go [hunk 33]
```diff
@@ -0,0 +1,321 @@
...
+func (t *shared) Read(p *common.Page) {
+	if !p.IsFreelistPage() {
+		panic(fmt.Sprintf("invalid freelist page: %d, page type is %s", p.Id(), p.Typ()))
+	}
+
+	ids := p.FreelistPageIds()
+
+	// Copy the list of page ids from the freelist.
+	if len(ids) == 0 {
+		t.Init(nil)
```
Comment (fuweid) [inline internal/freelist/shared.go:277]: For `array` type, the original implementation just resets `ids`. In this patch, `array` type will updates `t.cache` as well. It might involve some memory allocation but it looks fine.

File: internal/freelist/freelist.go [hunk 34]
```diff
@@ -0,0 +1,82 @@
...
+	Copyall(dst []common.Pgid)
+
+	// Reload reads the freelist from a page and filters out pending items.
+	Reload(p *common.Page)
+
+	// NoSyncReload reads the freelist from Pgids and filters out pending items.
+	NoSyncReload(pgIds common.Pgids)
+
+	// freePageIds returns the IDs of all free pages.
+	freePageIds() common.Pgids
```
Comment (fuweid) [inline internal/freelist/freelist.go:69]: These non-accessable methods could be considered to remove from `Interface` and we can add a new non-accessable interface for `array` type and `hash` type, like.

```go
type pageIDManager interface {
        freePageIds() common.Pgids
        pendingPageIds() map[common.Txid]*txPending
        release(txId common.Txid)
        releaseRange(begin, end common.Txid)
        mergeSpans(ids common.Pgids)
}
```

**Review by fuweid — APPROVED** [review]
/lgtm
/approve

File: internal/freelist/freelist.go (hunk 34 above)
Comment (ahrtr) [reply internal/freelist/freelist.go:69]: We can continue to refactor the interface later.

**Comment by ahrtr** [comment]
Link to https://github.com/etcd-io/bbolt/issues/789

File: internal/freelist/shared.go (hunk 33 above)
Comment (tjungblu) [reply internal/freelist/shared.go:277]: yep, I found some inconsistencies also during unit testing between the default implementations init (empty vs. nil). We'll fix them along with the tests.
